### DISCUSSION TITLE
Reimplementing checkio.org

### DISCUSSION THREAD

**Discussion by reznikmm** [alice-adventures/Alice#3] [General]
Before we lost Ada in Russian domain, I had a similar to Alice idea. I was attracted to the [checkio.org](https://checkio.org/) website (it's awesome!), so I tried to reimplement it for Ada. They have many "missions" for JS and Python. All these tasks are in open sources. ([An example](https://github.com/CheckiO-Missions/checkio-task-triangle-angles).) The tasks themselves are not very suitable for Ada, as they are designed for scripting languages, but 1) better that than nothing 2) many of them are translated into many languages, not just English, which seems important to me. If you're interested, I could try to recover my project, isolate it from the site code and package it in Docker, for example. I had about 8 missions ready, but that's enough for a demo. Of course I didn't use GNOGA.

**Comment by rocher** [comment]
This sounds great, Maxim! Didn't know checkio.org; looks amazing!

We could start with what you suggest, as a proof of concept. Later we can think in a better integration that facilitates and automates the import of missions and testing, if that makes sense. The advantage of checkio is that missions are open source and can be freely integrated (this is not the case of Project Euler and Advent of Code, for example).

Also, I think that this integration could bring the opportunity to create an environment where participants can explore and practice missions/exercises from a number of sources, like the couple of labs recently published in Alire.

So let's start with the first step. I created a repository in Alice for your work: [checkio-share](https://github.com/alice-adventures/checkio-share). Every problem source in Alice has at least a repository for shared content, <source>-shared. From it, participants will have access to a number of missions/exercises (your 8 missions, in this case). We will see later how to integrate it. We will probably need another repository to create a basic integration (what you have in your 'site code', but adapted for Alice).

Thanks Maxim for your initiative!!

**Comment by reznikmm** [comment]
Finally, I'm able to package it into a docker container. Sources and a very short instruction is here:

https://github.com/reznikmm/mimsy

Could you try it?

**Reply by rocher** [reply]
Good job Maxim!!  It really looks amazing and perfectly aligned with Alice ideas!!

Now we can think how to integrate it into Alice, and for this I think that there are two different paths.

First one, _the docker path_: you did a lot of work in the docker and there are lot of useful components inside it. Alice docker is going to look not so different, so I'm pretty sure it won't be not that difficult to integrate both.

Second one, _the Alire/GitHub path:_ all Developers and Participants should  be able to participate in any Problem Source with simple tools (editor, debugger, git client, etc) once they understand the underlying structure (crates/repositories).

Both paths will share a common resource: a repository with Checkio "missions". No matter if they decide to use the docker container or their preferred tools, they shall have access to all missions _ported_ to Alice (remember that I created the checkio-share repository for this).

I can help you with the second path because I'm doing something similar for Project Euler and other Problem Sources.

What's your view on this?

**Comment by reznikmm** [comment]
While the idea of chechio is indeed aligned with Alice, but it has a different approach, I guess. They put the engine on a single site, where users could solve missions and communicate each other by reading other's solutions, discussing them, put likes/dislikes on solutions/missions, follow users, so on. This is a kind of social network. And this could make a different experience than solving puzzles locally alone. On the other side, number of users for Ada probably will be very small, so a user stays alone any way :smile:.

I've tried your docker, but I still not quite understand how to port a mission to Alice. Could you explain or point some resources?
